### ISSUE TITLE
When switching to fuse2 my mounts show no directories with HNS account

### ISSUE THREAD

**Issue by jabbera** [kubernetes-sigs/blob-csi-driver#891]
**What happened**: Fuse 2 doesn't seem to be presenting directories. Before fuse2 my directory structure looked like this:

<img width="587" alt="image" src="https://user-images.githubusercontent.com/301208/234152968-13e03d74-3327-4f52-9052-f9c3ddccbcd4.png">

Now it looks like this:
<img width="294" alt="image" src="https://user-images.githubusercontent.com/301208/234150376-9b60906d-8275-4088-8422-e5f5edac4c1d.png">

All directories are missing!

**What you expected to happen**: Look like using fuse1

**How to reproduce it**: Mount my storage account in a pod!

**Anything else we need to know?**:
```
I0425 00:41:10.080466       1 nodeserver.go:185] mount inside driver with v2, protocol: fuse2, args: mount /var/lib/kubelet/pods/03770dd9-8519-4075-b591-bcd507464a48/volumes/kubernetes.io~csi/<SNIP>-prd-blob-storage/mount --cancel-list-on-mount-seconds=10 --file-cache-timeout-in-seconds=120 --log-level=LOG_WARNING -o allow_other -o attr_timeout=120 -o entry_timeout=120 -o negative_timeout=120 --pre-mount-validate=true --use-https=true --empty-dir-check=false --tmp-path=/mnt/csi-70dddfad53063ccf7c88b4b055b1b007e305cbae254858aca3e971d135822647 --container-name=prd
I0425 00:41:10.398767       1 nodeserver.go:599] blobfuse mount at /var/lib/kubelet/pods/03770dd9-8519-4075-b591-bcd507464a48/volumes/kubernetes.io~csi/<SNIP>-prd-blob-storage/mount success
I0425 00:41:10.398792       1 nodeserver.go:417] volume(csi-70dddfad53063ccf7c88b4b055b1b007e305cbae254858aca3e971d135822647) mount on "/var/lib/kubelet/pods/03770dd9-8519-4075-b591-bcd507464a48/volumes/kubernetes.io~csi/<SNIP>-prd-blob-storage/mount" succeeded
```
**Environment**: AKS 1.25.6

- CSI Driver version:  mcr.microsoft.com/oss/kubernetes-csi/blob-csi:v1.21.0
- Kubernetes version (use `kubectl version`): Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.6", GitCommit:"ff2c119726cc1f8926fb0585c74b25921e866a28", GitTreeState:"clean", BuildDate:"2023-01-30T18:18:41Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
- OS (e.g. from /etc/os-release): AKSCBLMariner-V2gen2-202304.10.0
- Kernel (e.g. `uname -a`): Whatever is on mariner
- Install tools: none
- Others: none

**Comment by andyzhangx** [comment]
@jabbera could you use AKS managed blob csi driver, follow this guide: https://learn.microsoft.com/en-us/azure/aks/azure-blob-csi?tabs=NFS#enable-csi-driver-on-a-new-or-existing-aks-cluster, your deployment does not have blobfuse-proxy, you could also install with `node.enableBlobfuseProxy=true` helm option.

**Comment by jabbera** [comment]
I'll have to wait a month until we have that upgrade scheduled.

**Comment by cvvz** [comment]
Could you share your workload manifest? Is this issue can 100% reproduce?

**Comment by jabbera** [comment]
> Can this issue be 100% reproduced?

@cvvz yes, 100% of the time

@andyzhangx does the proxy work on mariner? It's plastered all over the place that it's debian only.

Here is the requested data (but note the share mounts:

description
```
Name:                 csi-blob-node-2slgn
Namespace:            kube-system
Priority:             2000000000
Priority Class Name:  system-cluster-critical
Service Account:      csi-blob-node-sa
Node:                 aks-generalpool-40779947-vmss000003/10.15.40.151
Start Time:           Sun, 23 Apr 2023 22:49:33 -0400
Labels:               app=csi-blob-node
                      app.kubernetes.io/component=csi-driver
                      app.kubernetes.io/instance=blob-csi-driver
                      app.kubernetes.io/managed-by=Helm
                      app.kubernetes.io/name=blob-csi-driver
                      app.kubernetes.io/part-of=blob-csi-driver
                      app.kubernetes.io/version=v1.21.0
                      controller-revision-hash=c45fccf4b
                      helm.sh/chart=blob-csi-driver-v1.21.0
                      pod-template-generation=3
Annotations:          <none>
Status:               Running
IP:                   10.15.40.151
IPs:
  IP:           10.15.40.151
Controlled By:  DaemonSet/csi-blob-node
Containers:
  liveness-probe:
    Container ID:  containerd://13cfafdf31807e051b11535c75bbb482faa238c7ef00edb146fc33f4f84ded01
    Image:         mcr.microsoft.com/oss/kubernetes-csi/livenessprobe:v2.8.0
    Image ID:      mcr.microsoft.com/oss/kubernetes-csi/livenessprobe@sha256:fcb73e1939d9abeb2d1e1680b476a10a422a04a73ea5a65e64eec3fde1f2a5a1
    Port:          <none>
    Host Port:     <none>
    Args:
      --csi-address=/csi/csi.sock
      --probe-timeout=3s
      --health-port=29633
      --v=2
    State:          Running
      Started:      Sun, 23 Apr 2023 22:49:34 -0400
    Ready:          True
    Restart Count:  0
    Limits:
      memory:  100Mi
    Requests:
      cpu:     10m
      memory:  20Mi
    Environment:
    Mounts:
      /csi from socket-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-gwksp (ro)
  node-driver-registrar:
    Container ID:  containerd://43699a5ea19e60ff8d8f3affbe2f9cbff3fc1552a5c1c99182719227ce714180
    Image:         mcr.microsoft.com/oss/kubernetes-csi/csi-node-driver-registrar:v2.6.2
    Image ID:      mcr.microsoft.com/oss/kubernetes-csi/csi-node-driver-registrar@sha256:515b883deb0ae8d58eef60312f4d460ff8a3f52a2a5e487c94a8ebb2ca362720
    Port:          <none>
    Host Port:     <none>
    Args:
      --csi-address=$(ADDRESS)
      --kubelet-registration-path=$(DRIVER_REG_SOCK_PATH)
      --v=2
    State:          Running
      Started:      Sun, 23 Apr 2023 22:49:34 -0400
    Ready:          True
    Restart Count:  0
    Limits:
      memory:  100Mi
    Requests:
      cpu:     10m
      memory:  20Mi
    Liveness:  exec [/csi-node-driver-registrar --kubelet-registration-path=$(DRIVER_REG_SOCK_PATH) --mode=kubelet-registration-probe] delay=30s timeout=15s period=10s #success=1 #failure=3
    Environment:
      ADDRESS:                       /csi/csi.sock
      DRIVER_REG_SOCK_PATH:          /var/lib/kubelet/plugins/blob.csi.azure.com/csi.sock
    Mounts:
      /csi from socket-dir (rw)
      /registration from registration-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-gwksp (ro)
  blob:
    Container ID:  containerd://0b7e09a9647a68b3e01100da0c023aaee2c234c89d23d51cbd1229b98fa7fb4c
    Image:         mcr.microsoft.com/oss/kubernetes-csi/blob-csi:v1.21.0
    Image ID:      mcr.microsoft.com/oss/kubernetes-csi/blob-csi@sha256:ae6f80a20d10485ab6bcf1bac46129ef4919cc1186a03ecb4f623ba526e9de4d
    Port:          29633/TCP
    Host Port:     29633/TCP
    Args:
      --v=5
      --endpoint=$(CSI_ENDPOINT)
      --blobfuse-proxy-endpoint=$(BLOBFUSE_PROXY_ENDPOINT)
      --enable-blobfuse-proxy=false
      --nodeid=$(KUBE_NODE_NAME)
      --drivername=blob.csi.azure.com
      --cloud-config-secret-name=azure-cloud-provider
      --cloud-config-secret-namespace=kube-system
      --custom-user-agent=
      --user-agent-suffix=OSS-helm
      --allow-empty-cloud-config=true
      --enable-get-volume-stats=false
      --append-timestamp-cache-dir=false
      --mount-permissions=511
      --allow-inline-volume-key-access-with-idenitity=false
    State:          Running
      Started:      Sun, 23 Apr 2023 22:49:38 -0400
    Ready:          True
    Restart Count:  0
    Limits:
      memory:  2100Mi
    Requests:
      cpu:     10m
      memory:  20Mi
    Liveness:  http-get http://:healthz/healthz delay=30s timeout=10s period=30s #success=1 #failure=5
    Environment:
      CSI_ENDPOINT:                  unix:///csi/csi.sock
      KUBERNETES_PORT:               tcp://eastus2-gmo-research-cluster-nonprd.privatelink.eastus2.azmk8s.io:443
      KUBERNETES_PORT_443_TCP:       tcp://eastus2-gmo-research-cluster-nonprd.privatelink.eastus2.azmk8s.io:443
      AZURE_GO_SDK_LOG_LEVEL:
      KUBERNETES_PORT_443_TCP_ADDR:  eastus2-gmo-research-cluster-nonprd.privatelink.eastus2.azmk8s.io
      KUBERNETES_SERVICE_HOST:       eastus2-gmo-research-cluster-nonprd.privatelink.eastus2.azmk8s.io
      AZURE_CREDENTIAL_FILE:         <set to the key 'path' of config map 'azure-cred-file'>  Optional: true
      BLOBFUSE_PROXY_ENDPOINT:       unix:///csi/blobfuse-proxy.sock
      KUBE_NODE_NAME:                 (v1:spec.nodeName)
    Mounts:
      /csi from socket-dir (rw)
      /etc/kubernetes/ from azure-cred (rw)
      /mnt from blob-cache (rw)
      /var/lib/kubelet/ from mountpoint-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-gwksp (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  socket-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/plugins/blob.csi.azure.com
    HostPathType:  DirectoryOrCreate
  mountpoint-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/
    HostPathType:  DirectoryOrCreate
  registration-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/plugins_registry/
    HostPathType:  DirectoryOrCreate
  azure-cred:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/kubernetes/
    HostPathType:  DirectoryOrCreate
  blob-cache:
    Type:          HostPath (bare host directory volume)
    Path:          /mnt
    HostPathType:
  kube-api-access-gwksp:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.io/os=linux
Tolerations:                 op=Exists
Events:                      <none>
```

Log:
```
I0425 19:13:36.956049       1 utils.go:75] GRPC call: /csi.v1.Node/NodePublishVolume
I0425 19:13:36.956069       1 utils.go:76] GRPC request: {"target_path":"/var/lib/kubelet/pods/9ed3297e-584a-426f-874f-f7de729bf0f2/volumes/kubernetes.io~csi/snip-dev-blob-storage/mount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":7}},"volume_context":{"containerName":"dev","csi.storage.k8s.io/ephemeral":"true","csi.storage.k8s.io/pod.name":"mbtest","csi.storage.k8s.io/pod.namespace":"jhub","csi.storage.k8s.io/pod.uid":"9ed3297e-584a-426f-874f-f7de729bf0f2","csi.storage.k8s.io/serviceAccount.name":"default","mountOptions":"-o allow_other,--file-cache-timeout-in-seconds=120,--cancel-list-on-mount-seconds=10,-o attr_timeout=120,-o entry_timeout=120,-o negative_timeout=120,--log-level=LOG_DEBUG","protocol":"fuse2","secretName":"sas-snip-dev"},"volume_id":"csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352"}
I0425 19:13:36.956203       1 nodeserver.go:90] NodePublishVolume: ephemeral volume(csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352) mount on /var/lib/kubelet/pods/9ed3297e-584a-426f-874f-f7de729bf0f2/volumes/kubernetes.io~csi/snip-dev-blob-storage/mount, VolumeContext: map[containerName:dev csi.storage.k8s.io/ephemeral:true csi.storage.k8s.io/pod.name:mbtest csi.storage.k8s.io/pod.namespace:jhub csi.storage.k8s.io/pod.uid:9ed3297e-584a-426f-874f-f7de729bf0f2 csi.storage.k8s.io/serviceAccount.name:default getaccountkeyfromsecret:true mountOptions:-o allow_other,--file-cache-timeout-in-seconds=120,--cancel-list-on-mount-seconds=10,-o attr_timeout=120,-o entry_timeout=120,-o negative_timeout=120,--log-level=LOG_DEBUG protocol:fuse2 secretName:sas-snip-dev secretnamespace:jhub storageaccount:]
I0425 19:13:36.957286       1 blob.go:362] parsing volumeID(csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352) return with error: error parsing volume id: "csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352", should at least contain two #
I0425 19:13:36.957453       1 blob.go:427] volumeID(csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352) authEnv: []
I0425 19:13:36.995456       1 blob.go:782] got storage account(snip) from secret
I0425 19:13:36.995478       1 blob.go:512] accountSasToken is not empty, use it to access storage account(snip), container(dev)
I0425 19:13:36.995659       1 nodeserver.go:359] target /var/lib/kubelet/pods/9ed3297e-584a-426f-874f-f7de729bf0f2/volumes/kubernetes.io~csi/snip-dev-blob-storage/mount
protocol fuse2

volumeId csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352
context map[containerName:dev csi.storage.k8s.io/ephemeral:true csi.storage.k8s.io/pod.name:mbtest csi.storage.k8s.io/pod.namespace:jhub csi.storage.k8s.io/pod.uid:9ed3297e-584a-426f-874f-f7de729bf0f2 csi.storage.k8s.io/serviceAccount.name:default getaccountkeyfromsecret:true mountOptions:-o allow_other,--file-cache-timeout-in-seconds=120,--cancel-list-on-mount-seconds=10,-o attr_timeout=120,-o entry_timeout=120,-o negative_timeout=120,--log-level=LOG_DEBUG protocol:fuse2 secretName:sas-snip-dev secretnamespace:jhub storageaccount:]
mountflags []
mountOptions [--cancel-list-on-mount-seconds=10 --file-cache-timeout-in-seconds=120 --log-level=LOG_DEBUG -o allow_other -o attr_timeout=120 -o entry_timeout=120 -o negative_timeout=120 --pre-mount-validate=true --use-https=true --empty-dir-check=false --tmp-path=/mnt/csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352 --container-name=dev]
args /var/lib/kubelet/pods/9ed3297e-584a-426f-874f-f7de729bf0f2/volumes/kubernetes.io~csi/snip-dev-blob-storage/mount --cancel-list-on-mount-seconds=10 --file-cache-timeout-in-seconds=120 --log-level=LOG_DEBUG -o allow_other -o attr_timeout=120 -o entry_timeout=120 -o negative_timeout=120 --pre-mount-validate=true --use-https=true --empty-dir-check=false --tmp-path=/mnt/csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352 --container-name=dev
serverAddress snip.blob.core.windows.net
I0425 19:13:36.995818       1 nodeserver.go:185] mount inside driver with v2, protocol: fuse2, args: mount /var/lib/kubelet/pods/9ed3297e-584a-426f-874f-f7de729bf0f2/volumes/kubernetes.io~csi/snip-dev-blob-storage/mount --cancel-list-on-mount-seconds=10 --file-cache-timeout-in-seconds=120 --log-level=LOG_DEBUG -o allow_other -o attr_timeout=120 -o entry_timeout=120 -o negative_timeout=120 --pre-mount-validate=true --use-https=true --empty-dir-check=false --tmp-path=/mnt/csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352 --container-name=dev
I0425 19:13:37.179869       1 nodeserver.go:189] mount output:
I0425 19:13:37.203453       1 mount_linux.go:283] Detected umount with safe 'not mounted' behavior
I0425 19:13:37.223244       1 mount_linux.go:283] Detected umount with safe 'not mounted' behavior
I0425 19:13:37.244970       1 mount_linux.go:283] Detected umount with safe 'not mounted' behavior
I0425 19:13:37.263366       1 mount_linux.go:283] Detected umount with safe 'not mounted' behavior
I0425 19:13:37.289488       1 mount_linux.go:283] Detected umount with safe 'not mounted' behavior
I0425 19:13:37.289766       1 nodeserver.go:599] blobfuse mount at /var/lib/kubelet/pods/9ed3297e-584a-426f-874f-f7de729bf0f2/volumes/kubernetes.io~csi/snip-dev-blob-storage/mount success
I0425 19:13:37.289829       1 nodeserver.go:417] volume(csi-4b7f6caba7d95392f0ef0f9c28e4d36068b6103e61a716d3d640aed66bdd7352) mount on "/var/lib/kubelet/pods/9ed3297e-584a-426f-874f-f7de729bf0f2/volumes/kubernetes.io~csi/snip-dev-blob-storage/mount" succeeded
I0425 19:13:37.289869       1 utils.go:82] GRPC response: {}
```

Manifest:
```
apiVersion: v1
kind: Pod
metadata:
  name: mbtest
spec:
  automountServiceAccountToken: false
  containers:
  - command:
    - /bin/bash
    - -c
    - 'sleep 10000'
    image: ubuntu:latest
    name: mbtest
    securityContext:
      allowPrivilegeEscalation: true
      runAsUser: 0
    volumeMounts:
    - mountPath: /mnt/snip/dev
      name: snip-dev-blob-storage
  volumes:
  - csi:
      driver: blob.csi.azure.com
      volumeAttributes:
        containerName: dev
        mountOptions: -o allow_other,--file-cache-timeout-in-seconds=120,--cancel-list-on-mount-seconds=10,-o
          attr_timeout=120,-o entry_timeout=120,-o negative_timeout=120,--log-level=LOG_DEBUG
        protocol: fuse2
        secretName: sas-snip-dev
    name: snip-dev-blob-storage
```

**Comment by cvvz** [comment]
@jabbera  The logs looks all good. And I've tested from my side as well (**inline volume** with **sas token authentication** and **without blobfuse-proxy** and **mariner os** ), and everything was ok. 
![image](https://user-images.githubusercontent.com/44308864/234439585-1ef5fde2-a78a-424d-a8f7-4bed3e4beaa9.png)

Here are some actions you can take:

1. Make sure you have waited for about 10 seconds before executing `ls`. The mount option `--cancel-list-on-mount-seconds=10` means you need to wait for 10 seconds before you can `ls` the mount point after mounting successfully.

2. Try to mount manually on the mariner Node using blobfuse2 to see if the issue is still there.

**Comment by jabbera** [comment]
I'll try those things. One thing that may or may not make a difference is my storage account has HNS on.

**Comment by cvvz** [comment]
That's a useful information, let me try again.

**Comment by andyzhangx** [comment]
> @cvvz I'll try those things. One thing that may or may not make a difference is my storage account has HNS on.

@jabbera for HNS account, try add with "--use-adls=true" in the pv mountOptions
from AKS 1.25, the backend blobfuse mount would switch to blobfuse2 (v1 is not supported on Ubuntu 22.04 any more), I think there could be regression with HNS enabled account since v1 does not need this mount option. Will check with blobfuse team.

[Azure DataLake storage account](https://docs.microsoft.com/en-us/azure/storage/blobs/upgrade-to-data-lake-storage-gen2-how-to) support

set isHnsEnabled: "true" in storage class parameter to create ADLS account by driver in dynamic provisioning.
mount option --use-adls=true must be specified to enable blobfuse access ADLS account in static provisioning.

**Comment by jabbera** [comment]
@cvvz id you are trying again we use HNS and we are using userdelegationkeys for the SAS token.

**Comment by cvvz** [comment]
@jabbera I've tested and as Andy said, you need to add "--use-adls=true" to mountOptions for blobfuse2 and it would be ok. So there is a regression between blobfuse2 and blobfuse v1, while v1 can work without setting "--use-adls=true" explicitly.

**Comment by cvvz** [comment]
More specifically, without mounting with "--use-adls=true", you can't see all the directorys, and `mkdir` also return error.
![image](https://user-images.githubusercontent.com/44308864/234443372-ed86dfc0-14ce-4757-a740-ca7bccb27175.png)

**Comment by jabbera** [comment]
Thanks all. I'm confident this will resolve my issue! Will try tomorrow morning.

**Comment by jabbera** [comment]
Sorted! Thanks all.

**Comment by jabbera** [comment]
@cvvz I don't see any sign this was fixed. I'm trying to use RBAC since I have Storage Data Blob Contributor but I cannot because I have to specify --use-adls=true to get directories to show up. 

> there is a regression between blobfuse2 and blobfuse v1, while v1 can work without setting "--use-adls=true" explicitly.

**Comment by andyzhangx** [comment]
> @jabbera I've tested and as Andy said, you need to add "--use-adls=true" to mountOptions for blobfuse2 and it would be ok. So there is a regression between blobfuse2 and blobfuse v1, while v1 can work without setting "--use-adls=true" explicitly.

@vibhansa-msft do you think this is a regression between v1 & v2?
